### ISSUE TITLE
[RMM-15] Improve gradient clipping — clip gradients before accumulation

### ISSUE THREAD

**Issue by SkrOYC** [SkrOYC/langchain-middlewares-callbacks-ts#77]
**Parent:** #71 [RMM-9]
**Priority:** P3 — Training stability improvement
**Scope:** `src/middleware/hooks/after-model.ts`, `src/utils/matrix.ts`

## Problem

The current gradient clipping implementation clips **weights after update** instead of **gradients before accumulation**, and uses element-wise clipping instead of norm-based clipping.

**Current implementation (`matrix.ts:519-527`):**

```typescript
export function clipMatrix(matrix, minVal = -100, maxVal = 100) {
  return matrix.map(row => row.map(val => Math.max(minVal, Math.min(maxVal, val))));
}
```

**Usage (`after-model.ts`, weight update section):**

```typescript
const newWq = addMatrix(reranker.weights.queryTransform, gradWq);
const clippedWq = clipMatrix(newWq, -clipThreshold, clipThreshold);  // Clips WEIGHTS
```

**Issues:**

| Aspect | Current | Standard Practice |
|--------|---------|-------------------|
| Clipping target | Weights after update | Gradients before accumulation |
| Clipping method | Element-wise | L2 norm-based |
| Timing | After batch accumulation | Per-step, before accumulation |
| Threshold | Fixed ±100 | Often adaptive to dimension |

**Why this matters:**
- 4 samples accumulate before update — large accumulated gradients can exceed threshold before clipping applies
- Element-wise clipping distorts gradient direction (norm-based preserves it)
- Fixed ±100 doesn't scale with embedding dimension (1536-dim vs 768-dim)

### 1. Add norm-based gradient clipping utility

```typescript
export function clipMatrixByNorm(matrix: number[][], maxNorm: number): number[][] {
  const flat = matrix.flat();
  const norm = Math.sqrt(flat.reduce((sum, v) => sum + v * v, 0));
  if (norm <= maxNorm) return matrix;
  const scale = maxNorm / norm;
  return matrix.map(row => row.map(v => v * scale));
}
```

### 2. Apply clipping to gradients before accumulation

```typescript
// Before adding to accumulator
const clippedGradWq = clipMatrixByNorm(sampleGradients.gradWq, clipThreshold);
accumulator.accumulatedGradWq = addMatrix(
  accumulator.accumulatedGradWq,
  clippedGradWq
);
```

### 3. Keep weight clipping as a secondary safety net

The existing `clipMatrix` on weights can remain as a last-resort guard, but the primary clipping should happen on gradients.

## Acceptance Criteria

- [ ] New `clipMatrixByNorm(matrix, maxNorm)` utility in `matrix.ts`
- [ ] Gradient clipping applied before accumulation in `after-model.ts`
- [ ] Existing weight clipping retained as secondary guard
- [ ] All 559 existing tests pass
- [ ] New test: `clipMatrixByNorm` preserves direction when norm > threshold
- [ ] New test: `clipMatrixByNorm` returns identity when norm <= threshold
- [ ] New test: gradient accumulation respects clipping bounds
